controller/common/manage_children.go

```diff
@@ -126,6 +129,9 @@ func deleteChildren(client *dynamicclientset.ResourceClient, parent *unstructure
 func updateChildren(client *dynamicclientset.ResourceClient, updateStrategy ChildUpdateStrategy, parent *unstructured.Unstructured, observed, desired map[string]*unstructured.Unstructured) error {
 	var errs []error
 	for name, obj := range desired {
+		if parent.GetNamespace() == "" && obj.GetNamespace() != "" {
+			client = client.WithNamespace(obj.GetNamespace())
```

Comment:
Similar to above, I'd rather not assign back to `client`, since this is in a loop. However, unlike in `deleteChildren`, here we're working with desired objects returned by the hook, so we still need your `if` condition.

I suggest something like this to make it clear this is scoped to a single iteration of the loop:

```go
nsClient := client
if parent.GetNamespace() == "" && obj.GetNamespace() != "" {
  nsClient = nsClient.WithNamespace(obj.GetNamespace())
}
```

Then change `client` to `nsClient` in the rest of the function.

Reply:
Thanks for the code I think this should work. I worry that whoever the pour sole is that gets to rework the dynamic client will likely miss the subtle behavior here. We're relying on the fact that the client comes in with the namespace set to the parents namespace within these functions without actually checking that condition.

Reply:
Good point. What about something like this in each iteration?

```go
ns := obj.GetNamespace()
if ns == "" {
  ns = parent.GetNamespace()
}
nsClient := client.WithNamespace(ns)
```

Reply:
I'm testing that now, I should have the PR updated in a few minutes.

Reply:
It's been more than a few minutes but the tests passed. Sorry for the long wait. I think everything should be in order now.